### DISCUSSION TITLE
Correct way to run apps from Application List with required user permissions ?

### DISCUSSION THREAD

**Discussion by ivanmara555** [LizardByte/community#97] [Sunshine] [answered]
Hello! I am using sunshine on my windows 10 host installed with defaut installer "sunshine-windows.exe". Installer create sunshinesvc service running with NT AUTHORITY\SYSTEM permissions.

On shunshine windows host i have restricted user account in regular users group (not Administrators) where my desktop apps and some games.  When I need, i can switch to Administrator account or make "Run As Administartor". But always work as regular user.

Using moonlight client i can connect to sunshine windows host using default “Desktop” shortcut is hardcoded into sunshine. After connection i manually launch game from windows desktop shortcuts and game starting with correct regular user currently loggoned .

In sunshine web panel in Applications i create test game shortcut "Prodeus" with standart options like this:

> Application Name: Prodeus
> Command: D:\data\games\Prodeus\Prodeus.exe -gogEnabled
> Working Directory: D:\data\games\Prodeus
> Image: d:\data\games\Prodeus\prodeus.png
> 

After running test game shortcut "Prodeus" i found that game starts with NT AUTHORITY\SYSTEM user !!! 
Accordingly, all my settings and saves are unavailable, like you start game first time. Home folder for NT AUTHORITY\SYSTEM user become: %WinDir%\system32\config\systemprofile !

I think it's wrong to play games with  NT AUTHORITY\SYSTEM user account and store game settings in the system directory ... :)
And this is a big security problem.

**My question is: How i can correctly run apps from Application List with the required user rights ?**

p.s. Thanks for your working!

**Comment by ReenigneArcher** [comment]
Any command you run from sunshine will run with the same permissions of how sunshine is running.

Running sunshine as a service runs it as the `SYSTEM` user. From my understanding this is how all services are run. Running as admin will have similar results.

I have not tried this within sunshine, but you can try [psexec](https://learn.microsoft.com/en-us/sysinternals/downloads/psexec) to execute commands as another user.

**Comment by cgutman** [comment]
We should definitely be launching apps with active console user's token rather than our own SYSTEM token. Unfortunately, this means gutting the nice cross-platform Boost code.

There are a variety of ways to do this such as https://devblogs.microsoft.com/oldnewthing/20190425-00/?p=102443 (I _think_ this should work because sunshinesvc.exe spawns sunshine.exe into the user session, so I believe `GetShellWindow()` will work).

Rather than passing the shell process as a parent as in Raymond Chen's example code, we could also duplicate the token from it (like sunshinesvc itself does) and use `CreateProcessAsUserW()`.

**Reply by TheElixZammuto** [reply]
I tried the approach used by https://web.archive.org/web/20101009012531/http://blogs.msdn.com/b/winsdk/archive/2009/07/14/launching-an-interactive-process-from-windows-service-in-windows-vista-and-later.aspx (Method 1.A) and I can confirm it works with Sunshine, but I wasn't happy in how I built the feature and I scrapped it. I'm probably gonna try this again soon

**Comment by cgutman** [comment, marked as answer]
https://github.com/LizardByte/Sunshine/pull/600 will fix this when merged.

[a comment by an automated account is not shown]

**Reply by oliverw** [reply]
Is this supposed to be included/fixed in v0.18.4?

**Reply by ReenigneArcher** [reply]
It was v0.18.0 if I recall.
